[RFC][Doc] add a page describe actor execution order. (#23406)
ray-project/ray: doc/source/ray-core/actors/task-orders.rst

Actor Tasks Execution Orders
============================

Synchronous, Single-Threaded Actor
----------------------------------
In Ray, an actor receives tasks from multiple submitters (including driver and workers).

> For tasks received from the same submitter, synchronous, single-threaded actor executes

Reviewer: ```suggestion
For tasks received from the same submitter, a synchronous, single-threaded actor executes
```

them following the submission order.

> In other words, any task will not be executed until previously submitted tasks from

Reviewer: ```suggestion
In other words, a given task will not be executed until previously submitted tasks from
```

.. tabbed:: Python

    .. code-block:: python

        import time
        import ray

        @ray.remote
        class Counter:
            def __init__(self):
                self.value = 0

            def add(self, addition):
                self.value += addition
                return self.value

        counter = Counter.remote()

        # Submit task from a worker
        @ray.remote
        def submitter(value):
            return ray.get(counter.add.remote(value))

        # Simulate delayed result resolution.
        @ray.remote
        def delayed_resolution(value):
            time.sleep(5)
            return value

        # Submit tasks from different workers, with
        # the first submitted task waiting for
        # dependency resolution.
        value0 = submitter.remote(delayed_resolution.remote(1))
        value1 = submitter.remote(2)

        # Output: 3. The first submitted task is executed later.
        print(ray.get(value0))
        # Output: 2. The later submitted task is executed first.
        print(ray.get(value1))

Asynchronous or Threaded Actor
------------------------------

> :ref:`Asynchronous or threaded actors <async-actors>` does not guarantee the

Reviewer: ```suggestion
:ref:`Asynchronous or threaded actors <async-actors>` do not guarantee the
```

Actor Tasks Execution Orders
============================

Synchronous, single-threaded actors execute tasks according

> to their initial submission order. Any submitted task will not

Reviewer: Add some code examples? I think it's also worth noting that tasks can get reordered between different submitters. For example, if they are waiting on object argument resolution.

execute until previously submitted tasks have finished execution.
Note that actor task execution could fail. Upon failure, the following
tasks submitted to the same actor will not execute until the failed
task is successfully retried or the retry is exhausted.

:ref:`Async or threaded actors <async-actors>` does not guarantee the

> task execution order. For example, the system might execute a task

Reviewer: Didn't you add ways to configure this? Or is it always unordered?

Author: https://sourcegraph.com/github.com/ray-project/ray/-/blob/python/ray/_raylet.pyx?L1607 today it's always unordered.
